File: src/app/BluetoothTest/BluetoothTestApp/MainActivity.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Mime;
using System.Text;
using Android.App;
using Android.Content;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using Android.Bluetooth;
using Android.Bluetooth.LE;
using Android.OS;
using Java.Util;

namespace BluetoothTestApp
{
    [Activity(Label = "BluetoothTestApp", MainLauncher = true, Icon = "@drawable/icon")]
    public class MainActivity : Activity
    {
        int count = 1;
        private BluetoothSocket _socket;

        protected override void OnCreate(Bundle bundle)
        {
            base.OnCreate(bundle);

            // Set our view from the "main" layout resource
            SetContentView(Resource.Layout.Main);


            TextView Text = FindViewById<TextView>(Resource.Id.ResultText);
            Button SearchButton = FindViewById<Button>(Resource.Id.SearchButton);

            SearchButton.Click += (Object sender, EventArgs args) =>
            {
                BluetoothAdapter adapter = BluetoothAdapter.DefaultAdapter;
                if (adapter == null)
                    throw new Exception("No Bluetooth adapter found.");

                if (!adapter.IsEnabled)
                    throw new Exception("Bluetooth adapter is not enabled.");

                BluetoothDevice device = (from bd in adapter.BondedDevices
                                          where bd.Name == "DESKTOP-S7RVO4O"
                                          select bd).FirstOrDefault();

                if (device == null)
                    throw new Exception("Named device not found.");

                _socket = device.CreateRfcommSocketToServiceRecord(UUID.FromString("c3ffbcc2-ab89-4e56-94ed-2a8df65e45bd"));
                try
                {
                    _socket.Connect();
                }
                catch (Java.IO.IOException)
                {
                    AlertDialog.Builder builder = new AlertDialog.Builder(this);
                    AlertDialog alert = builder.Create();
                    alert.SetTitle("Disconnected");
                    alert.SetMessage("Do you want to reconnect?");

                    alert.SetButton("YES!", (s, ev) =>
                    {
                        
                    });
                    alert.SetButton("NO!", (s, ev) =>
                    {

                    });
                }

                if (_socket.IsConnected)
                    Text.SetText("Succes!".ToCharArray(), 0, 7);
                else
                    Text.SetText("FAIL!".ToCharArray(), 0, 5);

                // Read data from the device
                byte[] buffer = new byte[1024];

                try
                {
                    _socket.InputStream.Read(buffer, 0, buffer.Length);
                }
                catch (Java.IO.IOException)
                {
                    AlertDialog.Builder builder = new AlertDialog.Builder(this);
                    AlertDialog alert = builder.Create();
                    alert.SetTitle("Disconnected");
                    alert.SetMessage("Do you want to reconnect?");

                    alert.SetButton("YES!", (s, ev) =>
                    {

                    });
                    alert.SetButton2("NO!", (s, ev) =>
                    {

                    });
                    alert.Show();
                }

                string Message = Encoding.UTF8.GetString(buffer);
                TextView MessageText = FindViewById<TextView>(Resource.Id.MessageText);
                MessageText.SetText(Message.ToCharArray(), 0, Message.Length);


                // Write data to the device
                //_socket.OutputStream.Write(buffer, 0, buffer.Length);


            };
        }
    }
}

File: src/app/BluetoothTest/app/src/main/java/beuchert/bluetoothtest/MainActivity.java
package beuchert.bluetoothtest;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;
import android.content.DialogInterface;
import android.content.res.Resources;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.text.Editable;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.Set;
import java.util.UUID;

public class MainActivity extends AppCompatActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        TextView statusText = (TextView) findViewById(R.id.StatusText);
        Button connectButton = (Button) findViewById(R.id.ConnectButton);
        Button disconnectButton = (Button) findViewById(R.id.DisconnectButton);
    }

    // Fields:
    private String MacAdress = "00:0C:78:33:A5:63";
    private BluetoothAdapter BA = null;
    private BluetoothSocket mSocket = null;

    // onClick methods:
    public void connectClick(View view) {
        Connect();
    }

    public void DisconnectOnClick(View view){
        TextView statusText = (TextView) findViewById(R.id.StatusText);
        statusText.setText("Status: Not Connected");

        try {
            mSocket.close();
        } catch (IOException e) {
            ErrorMessage();
        }
    }

    public void ChangeOnClick(View view){
        EditText MacAdressEditText = (EditText) findViewById(R.id.NewMacText);
        Editable NewMacAdress = MacAdressEditText.getText();
        MacAdress = NewMacAdress.toString();
    }


    // Help methods:
    private void Connect(){
        final UUID MY_UUID = UUID.fromString("c3ffbcc2-ab89-4e56-94ed-2a8df65e45bd");

        BA = BluetoothAdapter.getDefaultAdapter();
        BluetoothDevice device = BA.getRemoteDevice(MacAdress);

        try {
            mSocket = device.createRfcommSocketToServiceRecord(MY_UUID);
            mSocket.connect();
        } catch (IOException e) {
            ErrorMessage();
        }

        TextView statusText = (TextView) findViewById(R.id.StatusText);
        statusText.setText("Connected to: " + device.getName());
        TextView PackageContent = (TextView) findViewById(R.id.PackageContent);

        if(mSocket.isConnected()){
            InputStream result = null;
            try {
                result = mSocket.getInputStream();
                BufferedReader r = new BufferedReader(new InputStreamReader(result));
                String realResult = r.readLine();
                PackageContent.setText(realResult);

            } catch (IOException e) {
                ErrorMessage();
            }
        }
    }

    private void ErrorMessage(){
        TextView statusText = (TextView) findViewById(R.id.StatusText);
        statusText.setText("Status: Not Connected");
        AlertDialog.Builder alertDiaglogBuilder = new AlertDialog.Builder(this);
        alertDiaglogBuilder.setTitle("Connection Error");
        alertDiaglogBuilder.setMessage("There was a bluetooth connection error (IOException). Do you want to try again?");
        alertDiaglogBuilder.setPositiveButton("Yes", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                Connect();
            }
        });

        alertDiaglogBuilder.setNegativeButton("No", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                try {
                    mSocket.close();
                } catch (IOException e) {
                    ErrorMessage();
                }
            }
        });

        AlertDialog alertDialog = alertDiaglogBuilder.create();
        alertDialog.show();
    }
}
